CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java

```diff
@@ -0,0 +1,513 @@
+			BackwardQuery query = new BackwardQuery(new Statement((Stmt) pred,sootMethod), new Val(providerValue, sootMethod));
+			//Submit query to the solver.
+			
+			BackwardBoomerangResults<NoWeight> backwardQueryResults = solver.solve(query);
+			map.putAll(backwardQueryResults.getAllocationSites());
+		}
+		
+		// The Provider can be correctly detected from this static analysis, if there is only one allocation site
+		// where the Provider variable was initialized. Otherwise, it throws an error because it is not possible
+		// to detect for sure the provider, if is given as parameter to the getInstance() method through the use of
+		// IF-ELSE, SWITCH statements or TERNARY operators
+		if(map.size() == 1) {
+			for(Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry : map.entrySet()) {
+				ForwardQuery forwardQuery = entry.getKey();
+				
+				Val forwardQueryVal = forwardQuery.var();
+				Value value = forwardQueryVal.value();
+				Type valueType = value.getType();
+				String valueTypeString = valueType.toString();
+				
+				// In here are listed all the supported providers so far
+				if(valueTypeString.contains("BouncyCastleProvider")) {
+					provider = "BC";
+				}
+			}
+		}
+		else if (map.size() > 1) {
+			throw new RuntimeException("The provider parameter must be passed directly to the"
+					+ " getInstance() method call, and not through IF-ELSE, SWITCH statements or"
+					+ " TERNARY operators.");
+		}
+		else {
+			throw new RuntimeException("Error occured to detect provider in the Provider Detection"
+					+ " analysis.");
+		}
+		return provider;
+	}
+	
+	
+	/**
+	 * This method return the provider used when Provider detected is of type `java.lang.String`
+	 * 
+	 * @param providerValue
+	 *            
+	 * @param body
+	 *            - i.e. the ActiveBody
+	 *            
+	 */
+	private String getProviderWhenTypeString(Value providerValue, Body body) {
+		for(Unit unit : body.getUnits()) {
+			if(unit instanceof JAssignStmt) {
+				JAssignStmt assignStatement = (JAssignStmt) unit;
+				if(assignStatement.getLeftOp().equals(providerValue)) {
+					return assignStatement.getRightOp().toString().replaceAll("\"","");
+				}
+			}
+		}
+		
+		return null;
+	}
+	
+	
+	/**
+	 * This method checks if the provider detected has only one allocation site
+	 * and it is not flowing through IF-ELSE statements or TERNARY operators, because
+	 * otherwise the provider can not be correctly detected through the use of
+	 * static analysis
+	 * 
+	 * @param providerValue
+	 *            
+	 * @param body
+	 *            - i.e. the ActiveBody
+	 *            
+	 */
+	private void checkIfStmt(Value providerValue, Body body) {
+		String value = providerValue.toString();
+		for(Unit unit : body.getUnits()) {
+			if(unit instanceof JIfStmt) {
+				JIfStmt ifStatement = (JIfStmt) unit;
+				if(ifStatement.toString().contains(value)) {
+					throw new RuntimeException("The provider parameter must be passed directly to the"
```

Comment:
All the runtime exception will crash our analysis. Log a warning and default to the JCA.